packages/workbox-background-sync/lib/Queue.mjs

```diff
@@ -0,0 +1,297 @@
+    } else {
+      names.add(name);
+    }
+
+    this._name = name;
+    this._callbacks = callbacks;
+    this._maxRetentionTime = maxRetentionTime;
+
+    this._addSyncListener();
+  }
+
+  /**
+   * Returns an object containing the `fetchDidFail` lifecycle method. This
+   * object can be used as a RequestWrapper plugin to automatically add
+   * failed requests to the background sync queue to be retried later.
+   *
+   * @return {Object}
+   */
+  createPlugin() {
+    return {
+      fetchDidFail: ({request}) => this.addRequest(request),
+    };
+  }
+
+
+  /**
+   * Stores the passed request into IndexedDB. The database used is
+   * `workbox-background-sync` and the object store name is the same as
+   * the name this instance was created with (to guarantee it's unique).
+   *
+   * @param {Request} request The request object to store.
+   */
+  async addRequest(request) {
+    const storableRequest = await StorableRequest.fromRequest(request);
+
+    this._runCallback('requestWillEnqueue', storableRequest);
+
+    const db = await this._getDb();
+    await db.add({
+      queueName: this._name,
+      storableRequest: storableRequest.toObject(),
+    });
+
+    // Schedule this, but don't await it as we don't want to block subsequent
+    // calls if the service worker isn't yet activated.
+    this._registerSync();
+  }
+
+  /**
+   * Retrieves all stored requests in IndexedDB and retries them. If the
+   * queue contained requests that were successfully replayed, the
+   * `allRequestsDidReplay` callback is invoked (which implies the queue is
+   * now empty). If any of the requests fail, a new sync registration is
+   * created to retry again later.
+   */
+  async replayRequests() {
+    const storableRequestsInQueue = await this._getStorableRequestsInQueue();
+
+    // If nothing is in the queue, return immediately and run no callbacks.
+    if (!storableRequestsInQueue.length) return;
+
+    const successfullyReplayedRequests = [];
+    let allReplaysSuccessful = true;
+
+    for (const [key, storableRequest] of storableRequestsInQueue) {
+      this._runCallback('requestWillReplay', storableRequest);
+
+      const request = storableRequest.toRequest();
+      let response;
+
+      try {
+        response = await fetch(request);
+      } catch (error) {
+        this._runCallback('replayDidFail', {error, request});
+        allReplaysSuccessful = false;
+        continue;
+      }
+
+      this._runCallback('requestDidReplay', {request, response});
+      successfullyReplayedRequests.push({request, response});
+
+      // Remove the request from IndexedDB asynchronously (don't await).
+      // TODO(philipwalton): in the unlikely event that the delete fails,
+      // this request may be replayed again. Do we want to handle this case?
+      this._removeRequest(key);
+    }
+
+    if (allReplaysSuccessful) {
+      this._runCallback('allRequestsDidReplay', successfullyReplayedRequests);
```

Comment:
I guess I could. Though if I did that I'm not sure why anyone would ever use the `replayDidFail` or `requestDidReplay` callbacks instead of this.

Maybe make this change and remove those two?